data/registry_model/quota.py

```diff
@@ -0,0 +1,353 @@
+    update_sizes(repository_id, manifest_id, storage_sizes, "subtract")
+
+
+def update_sizes(repository_id: int, manifest_id: int, storage_sizes, operation: str):
+    namespace_id = get_namespace_id_from_repository(repository_id)
+
+    # Addition - if the blob already referenced it's already been counted
+    # Subtraction - should only happen on the deletion of the last blob, if another exists
+    # don't subtract
+    namespace_total = 0
+    repository_total = 0
+    for storage_size in storage_sizes:
+        # To account for schema 1, which doesn't include the compressed_size field
+        blob_size = storage_size[IMAGE_SIZE] if storage_size[IMAGE_SIZE] is not None else 0
+        if not blob_exists_in_namespace(namespace_id, manifest_id, storage_size[ID]):
+            namespace_total = namespace_total + blob_size
+
+        if not blob_exists_in_repository(repository_id, manifest_id, storage_size[ID]):
+            repository_total = repository_total + blob_size
+
+    write_namespace_total(namespace_id, manifest_id, namespace_total, operation)
+    write_repository_total(repository_id, manifest_id, repository_total, operation)
+
+
+def blob_exists_in_namespace(namespace_id: int, manifest_id: int, blob_id: int):
+    try:
+        (
+            ManifestBlob.select(ManifestBlob.id)
+            .join(Repository, on=(ManifestBlob.repository == Repository.id))
+            .where(
+                Repository.namespace_user == namespace_id,
+                ManifestBlob.blob == blob_id,
+                ManifestBlob.manifest != manifest_id,
+            )
+            .get()
+        )
+        return True
+    except ManifestBlob.DoesNotExist:
+        return False
+
+
+def blob_exists_in_repository(repository_id: int, manifest_id: int, blob_id: int):
+    try:
+        (
+            ManifestBlob.select(ManifestBlob.id)
+            .where(
+                ManifestBlob.repository == repository_id,
+                ManifestBlob.blob == blob_id,
+                ManifestBlob.manifest != manifest_id,
+            )
+            .get()
+        )
+        return True
+    except ManifestBlob.DoesNotExist:
+        return False
+
+
+def write_namespace_total(
+    namespace_id: int, manifest_id: int, namespace_total: int, operation: str
+):
+    namespace_size = get_namespace_size(namespace_id)
+    namespace_size_exists = namespace_size is not None
+
+    # If backfill hasn't ran yet for this namespace don't do anything
+    if namespace_size_exists and (
+        namespace_size.backfill_start_ms is None
+        or namespace_size.backfill_start_ms > get_epoch_timestamp_ms()
+    ):
+        return
+
+    # If the namespacesize entry doesn't exist and this is the only manifest in the namespace
+    # we can assume this is the first push to the namespace and there is no blobs to be
+    # backfilled, so let the entry be created. Otherwise it still needs to be handled by the
+    # backfill worker so let's exit
```

Comment:
What's the lifecycle of a namespacesize? Does it get GC's along with the manifest? If so, then this assumption isn't valid anymore.